### ISSUE TITLE
Consider `httpResponse.request()` and deprecating `httpResponse.method()` to 

### ISSUE THREAD

**Issue by adriancole** [openzipkin/brave#1086]
In Brave 5.7,  `HttpClientRequest`, `HttpClientResponse`, `HttpServerRequest` and `HttpServerResponse`. These are wrappers for underlying types like HttpServletResponse.
Existing code needed to carry-over a use case of reading the HTTP method from a response callback. This was used to do things like make a route-based span name.

#1085  normalizes this to `HttpClientResponse` and `HttpServerResponse` commonality into `HttpResponse`. @anuraaga had an idea that we add `HttpResponse.request()` which could navigate to the corresponding method. `response.request()` would allow more flexible after-the-fact naming vs `response.method()`. For example, someone could fallback to a path based name. Rag may have other ideas, also.

There are a few things we should consider when doing this:

* `HttpClientResponse.method()` and `HttpServerResponse.method()` are already public, even if normalized, so we can't remove them.. however we could make them dispatch and deprecate.
* Sometimes the request reference changes between the start and finish stages of the span. This is particularly possible with some requests being immutable. We'd have to decide to always pass the same reference we parsed, or document that the request returned in `HttpResponse.request().unwrap()` may not be the same reference as the one we started with.
* There may be cases where we can't conveniently pass the original request directly to the response handler, and events may not have a full reference to the request. (thinking async servlet). This may imply response.request() is nullable.
* Holding a reference to a string for method is safer than holding a request object until a potentially async response/error callback. If we think this danger exists, we need `HttpResponse.request()` nullable to permit us flexibility to undo an unforeseen disaster.

Mostly, I'm concerned about the last point, and maybe @felixbarny @nicmunroe @raphw @eyalkoren know of some horror stories about holding references to requests during response callbacks.. I'll admit maybe I'm too nervous.

**Comment by nicmunroe** [comment]
Yeah there are certainly cases where you can get into trouble hanging onto request references, but it's not super common in my personal experience (YMMV depending on the services you tend you support). So I'm not sure how worried you should be...

One example I can think of - undertow will start cleaning up the request before it calls the async listeners for async requests: https://issues.redhat.com/browse/UNDERTOW-1437. Or at least it used to, I'm not sure if newer versions ever cleared that up. And that was for request attributes, not request method, but it's indicative of the kind of nastiness you can run into depending on the framework.

I think having `HttpResponse.request()` be nullable is sensible since you can't control the crazy things frameworks can and will do. Maybe you fallback to trying to grab the method via string if the request object comes back null? In that case you may not want to deprecate `response.method()`, but its purpose changes and it becomes the fallback rather than the primary?

**Comment by adriancole** [comment]
These are all good points @nicmunroe and appreciate the feedback. I thought I recalled some frameworks trying to recycle objects they gave users references to..

Regardless, I think it would have to be nullable anyway because when we retrofit our old adapter model.. well it never expected to have a reference to the request: we can't adapt existing libraries without allowing `Response.request()` to be null.

It is a good point about when you can't get a ref to the request, still passing method is helpful.. maybe it isn't useful to eagerly deprecate `Response.method()`.. particularly it is a primary input to the route-based name function.

Also, in revisiting an old issue #851, reading the servlet spec again.. it is clear we can make no promised about which request object would be returned via `Response.request()`, even in mutable frameworks. There's a higher likelihood that the request won't be predictable in an immutable framework, but the reverse isn't true.

TL;DR; is we can add a path to the request and do our best to provide it :) some baked-in parsers could reduce the work needed to double-check nulls.. and that's something we can solve separately.
